lib/dpl/provider/pages.rb

```diff
@@ -48,29 +57,70 @@ def check_auth
       end
 
       def needs_key?
+        # Provider's generic support for keys is about provisioning and
+        # deprovisioning them, while this provider needs to install an encrypted
+        # key provided by the caller.
         false
       end
 
+      def create_git_dir(gitdir)
+        if @gh_token
+          target_repo = "https://#{@gh_token}@#{@gh_ref}"
+        else
+          # Depend on the key set up in push_app.
+          target_repo = "git@github.com:#{slug}.git"
+        end
+        unless context.shell "git clone --branch '#{@target_branch}' --depth 1 '#{target_repo}' '#{gitdir}'"
+          FileUtils.mkdir(gitdir)
+          FileUtils.cd(gitdir) do
+            context.shell "git init" or error 'Could not create new git repo'
+            context.shell "git remote add origin '#{target_repo}'"
+            context.shell "git checkout -b '#{@target_branch}'"
+          end
+        end
+        FileUtils.cd(gitdir) do
+          context.shell "git config user.email '#{@gh_email}'"
+          context.shell "git config user.name '#{@gh_name}'"
+        end
+        return true
+      end
+
       def github_deploy
-        context.shell 'rm -rf .git > /dev/null 2>&1'
-        context.shell "touch \"deployed at `date` by #{@gh_name}\""
-        context.shell 'git init' or raise 'Could not create new git repo'
-        context.shell "git config user.email '#{@gh_email}'"
-        context.shell "git config user.name '#{@gh_name}'"
         context.shell "echo '#{@gh_fqdn}' > CNAME" if @gh_fqdn
         context.shell 'git add .'
+        if context.shell "git diff --cached --quiet"
```

Comment:
I'd prefer to give user more control over an empty commit https://github.com/webknjaz/dpl/commit/c8201a93163142dc43245ec5c097d6a054307245, having reasonable default, of course.

Reply:
That's plausible, but new options have a cost, and my impression is that @BanzaiMan doesn't want this one. I'm happy to add this option if they prefer to have it. See also https://github.com/travis-ci/dpl/pull/719#pullrequestreview-79236861.